### ISSUE TITLE
detect target platform consistently, and simplify cross compilation

### ISSUE THREAD

**Issue by katjav** [pure-data/pd-lib-builder#55]
Experimental branch `target-detection` has a series of commits (starting with `1dc4219b`) to implement consistent target platform detection (hardware architecture and OS). It is based on target triplet as reported by compiler with option `-dumpmachine`. This option was first suggested by @umlaute in the context of issue #37, to detect architectures in the case of Windows. 

The jungle of GNU triplets scared me initially (some are even quadruplets), but of course pdlibbuilder supports only a small subset of them. For our purposes, the only distinction that triplets can't make is between `armv6` and `armv7` (both `arm-linux-gnueabihf`). For `armv6` (first gen RPi, RPi Zero) I stayed with autodetection in native compilation only, and it must be explicited otherwise.

Commit `262143b` introduces an optional user variable `PLATFORM` to hold a target triplet which is parsed as the base name for tools required for the build (preprocessor, compilers and strip). If the tool chain is available in the path, no other variables need to be overriden for cross compilation. It is the equivalent of autotools' `--host` parameter.

Using variable `PLATFORM` I could cross-compile for Linux (Intel and ARM), OSX and Windows, on a single Linux 64 bit machine with cross tool chains installed for all those platforms. Native compilation must be tested as well to make sure no regressions are introduced. I have no access to the following platforms: OSX > 10.8, Windows 64 and 32 bit. @lucarda, @danomatika, could you give branch `target-detection` a try?

Oh I almost forgot to mention: fat-binary-default for OSX was sacrified in the process. It felt so stupid to poll the compiler for its target architecture and still add another architecture. See commit `65e9fc7` for a user-friendly alternative to get fat builds.

**Comment by danomatika** [comment]
>fat-binary-default for OSX was sacrified in the process

That's fine as macOS 10.14+ cannot build for 32 bit and will throw an error anyway.

**Comment by Lucarda** [comment]
> could you give branch `target-detection` a try?

All good in Windows with Msys2 (64 & 32).

**Comment by katjav** [comment]
> All good in Windows with Msys2 (64 & 32).

Happy to hear, thanks for testing. If MacOS is fine too I'll merge this.

**Comment by Lucarda** [comment]
> If MacOS is fine too I'll merge this.

May be also @porres can test this on macOS ?

**Comment by electrickery** [comment]
Native compilation and execution is without problems with this branch 
version (f2a5b66 
<https://github.com/pure-data/pd-lib-builder/commit/f2a5b6614cb6ae9231326f4d44a9508af46d27b1>). 
Tested with the latest MacOSX 10.11.6 on a MacMini (this machine is 10 
years old by now). Tested with the latest pd-nilwind and pd-maxlib.

Greetings,

Fred Jan

**Comment by katjav** [comment]
> Tested with the latest MacOSX 10.11.6 on a MacMini

OK thanks for testing. Under the assumption they didn't drop option `-dumpmachine` since 10.11 it seems safe to merge this target detection stuff.

**Comment by danomatika** [comment]
On macOS 10.14.6:
~~~
$ make dumpmachine
++++ info: using Makefile.pdlibbuilder version 0.5.1
++++ info: making target dumpmachine in lib helloworld
x86_64-apple-darwin18.7.0
~~~

I tested with the helloworld example and it builds and loads fine for 64 bit.

**Comment by danomatika** [comment]
Also, I suggest opening a PR of the test branch into the master branch. Then the diffs are easy to see and review within the frame of the discussion.

**Comment by katjav** [comment]
> Also, I suggest opening a PR of the test branch into the master branch. Then the diffs are easy to see and review within the frame of the discussion.

Thanks for testing and for the tip. I'll first document variable `PLATFORM` etc. to make this topic complete, then do a PR.

**Comment by katjav** [comment]
Branch merged and version updated to v0.6.0 for this.
